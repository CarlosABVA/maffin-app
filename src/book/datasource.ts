import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
