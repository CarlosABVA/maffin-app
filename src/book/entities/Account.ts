import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Entity, JoinColumn,
  ManyToOne,
  Tree,
  TreeParent,
  TreeChildren,
  OneToMany, PrimaryColumn,
} from 'typeorm';

import type Commodity from './Commodity';
import Money from '../Money';
import type Split from './Split';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static TYPES = ['ROOT', 'ASSET', 'BANK', 'CASH', 'EQUITY', 'LIABILITY', 'INCOME', 'EXPENSE', 'MUTUAL', 'STOCK'];

  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.MaxLength(2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @v.IsIn(Account.TYPES)
    type!: string;

  @Column({
    type: 'text',
    select: false,
    default: '',
  })
    path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @TreeChildren()
    children!: Account[];

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'commodity_guid' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  /**
   * Note this only works if the entity has access to splits
   */
  get total(): Money {
    if (this.type === 'ROOT') {
      // @ts-ignore
      return null;
    }

    const total = this.splits.reduce(
      (acc, split) => acc.add(
        new Money(split.quantityNum / split.quantityDenom, this.commodity.mnemonic),
      ),
      new Money(0, this.commodity.mnemonic),
    );

    if (total.isNegative()) {
      return total.multiply(-1);
    }

    return total;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });
