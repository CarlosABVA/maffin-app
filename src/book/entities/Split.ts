import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Relation,
  Entity, JoinColumn, ManyToOne, PrimaryColumn,
} from 'typeorm';

import type Account from './Account';
import type Transaction from './Transaction';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE splits (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   tx_guid         CHAR(32) NOT NULL,
 *   account_guid    CHAR(32) NOT NULL,
 *   memo            text(2048) NOT NULL,
 *   action          text(2048) NOT NULL,
 *   reconcile_state text(1) NOT NULL,
 *   reconcile_date  timestamp NOT NULL,
 *   value_num       integer NOT NULL,
 *   value_denom     integer NOT NULL,
 *   quantity_num    integer NOT NULL,
 *   quantity_denom  integer NOT NULL,
 *   lot_guid        CHAR(32)
 * );
*/
@Entity('splits')
export default class Split extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @ManyToOne('Transaction')
  @JoinColumn({ name: 'tx_guid' })
    fk_transaction!: Relation<Transaction> | string;

  get transaction(): Transaction {
    return this.fk_transaction as Transaction;
  }

  @ManyToOne('Account')
  @JoinColumn({ name: 'account_guid' })
    fk_account!: Relation<Account> | string;

  get account(): Account {
    return this.fk_account as Account;
  }

  // For stock operations, action is set to Buy
  @PrimaryColumn({
    type: 'varchar',
    length: 2048,
  })
    action?: string = '';

  @Column({
    type: 'integer',
    name: 'value_num',
  })
    valueNum!: number;

  @Column({
    type: 'integer',
    name: 'value_denom',
  })
    valueDenom!: number;

  @Column({
    type: 'integer',
    name: 'quantity_num',
  })
    quantityNum!: number;

  @Column({
    type: 'integer',
    name: 'quantity_denom',
  })
    quantityDenom!: number;

  /**
   * Returns the value of the transaction in the transaction's currency
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR. This means you spent 1000 EUR to buy
   * 122.8501 of whatever the split's account currency is.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - value in transaction's currency
   */
  get value(): number {
    return this.valueNum / this.valueDenom;
  }

  /**
   * Returns the quantity spent in the currency of the split's account.
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR and the symbol of the account
   * is GOOGL. This means that you spent 1000 EUR to purchase 122.8501 GOOGL
   * stocks.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - quantity spent in purchasing in the split account's currency
   */
  get quantity(): number {
    return this.quantityNum / this.quantityDenom;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Split, 'name', { value: 'Split' });
