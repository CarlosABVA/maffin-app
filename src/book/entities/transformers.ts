import { DateTime } from 'luxon';
import { ValueTransformer } from 'typeorm';

export class DateTimeTransformer implements ValueTransformer {
  // eslint-disable-next-line class-methods-use-this
  from(value: string | null): DateTime | null {
    // For some reason, when using TreeRepository this is null at some point
    if (!value) {
      return null;
    }
    const result = DateTime.fromSQL(value);
    if (!result.isValid) {
      throw new Error(`invalid date: ${result.invalidExplanation}`);
    }
    return result;
  }

  // eslint-disable-next-line class-methods-use-this
  to(value: DateTime): string {
    const result = value.toSQLDate();
    if (result === null) {
      throw new Error(`invalid date: ${value.invalidExplanation}`);
    }
    return result;
  }
}
