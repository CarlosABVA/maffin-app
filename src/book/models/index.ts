export { default as InvestmentAccount } from './InvestmentAccount';
