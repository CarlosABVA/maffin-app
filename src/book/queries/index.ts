export { getInvestments } from './getInvestments';
export { getAccountsWithPath } from './getAccountsWithPath';
export { getMainCurrency } from './getMainCurrency';
