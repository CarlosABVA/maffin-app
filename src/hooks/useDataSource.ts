import React from 'react';
import { DataSource } from 'typeorm';

import useBookStorage from '@/hooks/useBookStorage';
import { initDB } from '@/book/datasource';

// Not sure if this is the best way to cache the storage...
// need to research
let DATASOURCE: DataSource | null = null;

export default function useDataSource(): [DataSource | null] {
  const [bookStorage] = useBookStorage();
  const [, setDataSource] = React.useState<DataSource | null>(DATASOURCE);
  const isInitializing = React.useRef(false);

  React.useEffect(() => {
    async function load() {
      if (bookStorage && DATASOURCE == null && !isInitializing.current) {
        isInitializing.current = true;
        let start;
        let end;

        start = performance.now();
        const rawBook = await bookStorage.get();
        end = performance.now();

        start = performance.now();
        // We re-check here to avoid race conditions because of
        // downloading book data takes time. Without this
        // we have some reference errors where a component may
        // use one instance another the difference instance which
        // causes synchronization issues.
        if (DATASOURCE === null) {
          DATASOURCE = await initDB(rawBook);
          end = performance.now();
          console.log(`init dataSource: ${end - start}ms`);
        }
        setDataSource(DATASOURCE);
        isInitializing.current = false;
      }
    }

    load();
  }, [bookStorage]);

  return [DATASOURCE];
}
