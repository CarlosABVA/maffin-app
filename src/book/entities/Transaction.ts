import { DateTime } from 'luxon';
import {
  BaseEntity,
  Column, CreateDateColumn, Entity, JoinColumn, ManyToOne, OneToMany,
  PrimaryColumn,
} from 'typeorm';
import * as v from 'class-validator';

import type Commodity from './Commodity';
import type Split from './Split';
import { DateTimeTransformer } from './transformers';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE transactions (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   currency_guid   CHAR(32) NOT NULL,
 *   num             text(2048) NOT NULL,
 *   post_date       timestamp NOT NULL,
 *   enter_date      timestamp NOT NULL,
 *   description     text(2048)
 * );
 */

@Entity('transactions')
export default class Transaction extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
    guid!: string;

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'currency_guid' })
    fk_currency!: Commodity | string;

  get currency(): Commodity {
    return this.fk_currency as Commodity;
  }

  @Column({
    type: 'date',
    transformer: new DateTimeTransformer(),
    name: 'post_date',
  })
    date!: DateTime;

  @CreateDateColumn({
    name: 'enter_date',
  })
    enterDate?: Date;

  @OneToMany('Split', (split: Split) => split.fk_transaction)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    default: '',
  })
  @v.MaxLength(2048)
    description!: string;
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Transaction, 'name', { value: 'Transaction' });
