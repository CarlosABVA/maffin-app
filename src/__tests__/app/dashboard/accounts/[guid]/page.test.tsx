import React from 'react';
import {
  act,
  render,
  screen,
} from '@testing-library/react';
import type { DataSource } from 'typeorm';

import AccountPage from '@/app/dashboard/accounts/[guid]/page';
import { TransactionsTableProps } from '@/components/TransactionsTable';
import { Account } from '@/book/entities';
import * as queries from '@/book/queries';
import * as dataSourceHooks from '@/hooks/useDataSource';

jest.mock('@/components/TransactionsTable', () => {
  function TransactionsTable(props: TransactionsTableProps) {
    return (
      <div className="TransactionsTable">
        <span>{JSON.stringify(props)}</span>
      </div>
    );
  }

  return TransactionsTable;
});

jest.mock('@/book/queries', () => ({
  __esModule: true,
  ...jest.requireActual('@/book/queries'),
}));

jest.mock('@/hooks/useDataSource', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useDataSource'),
}));

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}));
// eslint-disable-next-line @typescript-eslint/no-var-requires
const useRouter = jest.spyOn(require('next/navigation'), 'useRouter');

describe('AccountPage', () => {
  let mockRouterPush: jest.Mock;

  beforeEach(() => {
    mockRouterPush = jest.fn();
    useRouter.mockImplementation(() => ({
      push: mockRouterPush,
    }));
    jest.spyOn(queries, 'getAccountsWithPath').mockResolvedValue([
      {
        guid: 'guid',
        path: 'path',
        type: 'TYPE',
      } as Account,
    ]);
  });

  it('displays loading while account is null', () => {
    jest.spyOn(Account, 'findOneBy').mockResolvedValue(null);
    const { container } = render(<AccountPage params={{ guid: 'guid' }} />);

    expect(screen.getByText('Loading...')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('returns 404 when account not found', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{} as DataSource]);
    jest.spyOn(Account, 'findOneBy').mockResolvedValue(null);

    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      render(
        <AccountPage params={{ guid: 'guid' }} />,
      );
    });

    expect(mockRouterPush).toHaveBeenCalledWith('/404');
  });

  it('renders as expected with account', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{} as DataSource]);
    jest.spyOn(Account, 'findOneBy').mockResolvedValue({
      guid: 'accountGuid',
      name: 'Account name',
      path: 'account:name',
      commodity: {
        mnemonic: 'EUR',
      },
    } as Account);

    let container;
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      ({ container } = render(
        <AccountPage params={{ guid: 'guid' }} />,
      ));
    });

    expect(container).toMatchSnapshot();
  });
});
