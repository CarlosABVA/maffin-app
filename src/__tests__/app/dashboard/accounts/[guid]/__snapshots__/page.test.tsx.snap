// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountPage displays loading while account is null 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`AccountPage renders as expected with account 1`] = `
<div>
  <div
    class="grid grid-cols-12 items-center pb-4"
  >
    <span
      class="col-span-10 text-xl font-medium"
    >
      account:name
       
      account
    </span>
    <div
      class="col-span-2 col-end-13 justify-self-end"
    >
      <div
        class="fixed hidden inset-0 bg-white bg-opacity-30 overflow-y-auto h-full w-full z-50"
        id="my-modal"
      >
        <div
          class="relative top-20 mx-auto p-5 w-1/3 shadow-lg rounded-md bg-gunmetal-700"
        >
          <button
            class="float-right"
            type="button"
          >
            X
          </button>
          <div
            class="w-full pb-4 border-b border-gunmetal-600"
          >
            <span
              class="text-lg"
            >
              account:name transaction
            </span>
          </div>
          <form>
            <fieldset
              class="text-sm my-5"
            >
              <label
                class="inline-block mb-2"
                for="dateInput"
              >
                Date
              </label>
              <input
                class="block w-full m-0"
                id="dateInput"
                name="date"
                type="date"
              />
              <p
                class="invalid-feedback"
              />
            </fieldset>
            <fieldset
              class="text-sm my-5"
            >
              <label
                class="inline-block mb-2"
                for="descriptionInput"
              >
                Description
              </label>
              <input
                class="block w-full m-0"
                id="descriptionInput"
                name="description"
                placeholder="Enter description"
                type="text"
              />
              <p
                class="invalid-feedback"
              />
            </fieldset>
            <fieldset
              class="text-sm my-5"
            >
              <label
                class="inline-block mb-2"
              >
                Entries
              </label>
              <fieldset
                class="grid grid-cols-12"
              >
                <div
                  class="col-span-8"
                >
                  <div
                    class=" css-b62m3t-container"
                    id="splits.0.toAccount"
                  >
                    <span
                      class="css-1f43avz-a11yText-A11yText"
                      id="react-select-2-live-region"
                    />
                    <span
                      aria-atomic="false"
                      aria-live="polite"
                      aria-relevant="additions text"
                      class="css-1f43avz-a11yText-A11yText"
                    />
                    <div
                      class="!h-9 !min-h-fit text-sm pl-10 !bg-gunmetal-800 !border-none css-13cymwt-control"
                    >
                      <svg
                        class="absolute text-md left-[10px]"
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
                        />
                      </svg>
                      <div
                        class=" css-1fdsijx-ValueContainer"
                      >
                        <div
                          class=" css-1jqq78o-placeholder"
                          id="react-select-2-placeholder"
                        >
                          &lt;account&gt;
                        </div>
                        <div
                          class="!text-inherit css-qbdosj-Input"
                          data-value=""
                        >
                          <input
                            aria-autocomplete="list"
                            aria-describedby="react-select-2-placeholder"
                            aria-expanded="false"
                            aria-haspopup="true"
                            aria-label="splits.0.toAccount"
                            autocapitalize="none"
                            autocomplete="off"
                            autocorrect="off"
                            class=""
                            id="react-select-2-input"
                            role="combobox"
                            spellcheck="false"
                            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      <div
                        class=" css-1hb7zxy-IndicatorsContainer"
                      >
                        <span
                          class="hidden css-1u9des2-indicatorSeparator"
                        />
                        <div
                          aria-hidden="true"
                          class=" css-1xc3v61-indicatorContainer"
                        >
                          <svg
                            aria-hidden="true"
                            class="css-tj5bde-Svg"
                            focusable="false"
                            height="20"
                            viewBox="0 0 20 20"
                            width="20"
                          >
                            <path
                              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                            />
                          </svg>
                        </div>
                      </div>
                    </div>
                    <input
                      name="splits.0.toAccount"
                      type="hidden"
                      value=""
                    />
                  </div>
                  <p
                    class="invalid-feedback"
                  />
                </div>
                <div
                  class="col-span-2"
                >
                  <input
                    class="w-full text-right bg-gunmetal-800 border-none rounded-sm m-0"
                    name="splits.0.amount"
                    placeholder="0.0"
                    step="0.001"
                    type="number"
                  />
                  <p
                    class="invalid-feedback"
                  />
                </div>
                <div
                  class="col-span-1"
                >
                  <input
                    class="w-full text-center bg-gunmetal-700 border-none rounded-sm m-0"
                    disabled=""
                    value="€"
                  />
                </div>
                <div
                  class="col-span-1"
                >
                  <button
                    class="btn-primary rounded-sm"
                    type="button"
                  >
                    X
                  </button>
                </div>
              </fieldset>
              <button
                class="link m-3"
                type="button"
              >
                Add split
              </button>
            </fieldset>
            <div
              class="flex w-full gap-2 items-center justify-center"
            >
              <span
                class="block py-2 px-4 rounded-lg bg-gunmetal-800 text-sm"
              >
                0.00 €
              </span>
              <button
                class="btn-primary"
                type="submit"
              >
                Save
              </button>
            </div>
          </form>
        </div>
      </div>
      <div>
        <button
          class="btn-primary"
          type="button"
        >
          <svg
            class="mr-1"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4z"
            />
            <path
              d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
            />
          </svg>
          Add Transaction
        </button>
      </div>
    </div>
  </div>
  <div
    class="TransactionsTable"
  >
    <span>
      {"accountId":"accountGuid","accounts":[{"guid":"guid","path":"path","type":"TYPE"}]}
    </span>
  </div>
</div>
`;
